app/models/manageiq/providers/redfish/physical_infra_manager/physical_server/provisioning.rb

```diff
@@ -0,0 +1,42 @@
+module ManageIQ::Providers::Redfish
+  module PhysicalInfraManager::PhysicalServer::Provisioning
+    def deploy_pxe_config(pxe_image, customization_template)
+      with_provider_object do |system|
+        if (macs = mac_addresses(system)).empty?
+          raise MiqException::MiqProvisionError, 'at least one MAC address is needed for provisioning'
+        end
+        macs.each { |mac| pxe_image.pxe_server.create_provisioning_files(pxe_image, mac, nil, customization_template) }
+      end
+    end
+
+    def reboot_using_pxe
+      with_provider_object do |system|
+        response = system.patch(
+          :payload => {
+            'Boot' => {
+              'BootSourceOverrideEnabled' => 'Once',
+              'BootSourceOverrideTarget'  => 'Pxe'
+            }
+          }
+        )
+        raise MiqException::MiqProvisionError, 'Cannot override boot order' if response.status >= 400
+      end
+      # TODO: we perform force reboot which will fail in some cases. Need to handle with supports mixin.
+      restart_now
+    end
+
+    def powered_on_now?
+      # TODO(miha-plesko): we should rely on VMDB state instead contacting provider.
+      # Update implementation once we have event-driven targeted refresh implemented.
+      with_provider_object { |system| return system.PowerState.to_s.downcase == 'on' }
+    end
+
+    private
+
+    def mac_addresses(system)
+      system.EthernetInterfaces.Members.reduce(Set.new) do |acc, el|
+        acc.add(el.PermanentMACAddress).add(el.MACAddress)
```

Comment:
A few days ago I learned the hard way that not all ethernet interfaces have both of those fields set. Maybe we can filter out the `nil`s with something similar to this:

    acc.merge([el.PermanentMACAddress, el.MACAddress].compact)

Reply:
Thanks 👍 I've updated commit to properly handle even empty string values now, just in case some server thinks its a good joke to send them over.

Reply:
I think this method could be simplified a bit to be easier to read, here's what I'm thinking let me know if it matches how this currently works:

```ruby
system.EthernetInterfaces.Members.map { |iface| iface.MACAddress || iface.PermanentMACAddress }.compact
```

Reply:
> I've updated commit to properly handle even empty string values now, just in case some server thinks its a good joke to send them over.

@miha-plesko did you forgot to push the changes?

As for the `mac_addresses` method, its task is to collect all distinct MAC addresses from the Ethernet interfaces that are present on the system. For example, let us assume that we have four interfaces on our computer that report this back via Redfish API:

    eth0:
      MACAddress: mac1
      PermanentMACAddress: mac2
    eth1:
      MACAddress: mac3
    eth2:
      PermanentMACAddress: mac4
    eth3:
      MACAddress: mac5
      PermanentMACAddress: mac5

What we would like to get back is some iterable thing that contains all five MACs (mac1, mac2, ..., mac5). If we do naive `Members.flat_map { |iface| [iface.MACAddress, iface.PermanentMACAddress] }`, we will get back `["mac1", "mac2", "mac3", nil, nil, "mac4", "mac5", "mac5"]`. Now we need to filter out the `nil`s and remove duplicates. Calling `compact.to_set` on the `flat_map`'s return value would do the job, but I am not sure what would the most "rubyish" approach be.

Reply:
Oh okay I thought only one or the other could be set, sounds like `(Members.map(&:MACAddress) + Members.map(&:PermanentMACAddress)).reject(&:blank?)` is what you want then

Reply:
@agrare we'd also need to remove the duplicates because as far as I know MAC and PermanenctMAC are usually the same (fyi: MAC is hardware MAC while PermanentMAC is optionally spoofed mac addr). So in the end we'd have

```ruby
def mac_addresses(system)
  [].tap do |macs|
    macs.concat(system.EthernetInterfaces.Members.map(&:MACAddress))
    macs.concat(system.EthernetInterfaces.Members.map(&:PermanentMACAddress))
    macs.reject(&:blank?).uniq  # get rid of nils, empty strings and duplicates
  end
end
```

Wdyt? I like both methods and can agree that this second one is more obvious.

@tadeboro the change I mentioned there is in L39 which removes all nils and emtpy strings from the set.

Reply:
@miha-plesko you can just tack a `.uniq` to the end of mine if you're worried about duplicate macs

Reply:
Agree with @agrare suggestion, this is the more readable -
`system.EthernetInterfaces.Members.map { |iface| iface.MACAddress.presence || iface.PermanentMACAddress.presence }.compact.uniq`

Reply:
I really do not care what implementation we use, as long as this invariant holds true:

    require "redfish_client/resource"
    system = RedfishClient::Resource.new(nil, content: {
      "EthernetInterfaces" => {
        "Members" => [
          { "MACAddress" => "mac1", "PermanentMACAddress" => "mac2" },
          { "MACAddress" => "mac3"                                  },
          {                         "PermanentMACAddress" => "mac4" },
          { "MACAddress" => "mac5", "PermanentMACAddress" => "mac5" },
        ]
      }
    })
    mac_addresses(system).to_set == Set.new(%w[mac1 mac2 mac3 mac4 mac5])